File: apps/sdkwork-rtc-flutter-mobile/lib/bootstrap/host_adapters.dart
typedef RtcMediaPermissionAdapter = Future<bool> Function();

class RtcDeepLinkAdapter {
  const RtcDeepLinkAdapter({
    required this.currentPath,
    required this.subscribe,
  });

  final String Function() currentPath;
  final void Function(void Function(String path) listener) subscribe;
}

class RtcSecureStorageAdapter {
  const RtcSecureStorageAdapter({
    required this.read,
    required this.write,
    required this.remove,
  });

  final String? Function(String key) read;
  final void Function(String key, String value) write;
  final void Function(String key) remove;
}

class RtcHostAdapters {
  const RtcHostAdapters({
    this.camera,
    this.microphone,
    this.deepLinks,
    this.pushNotifications,
    this.secureStorage,
  });

  final RtcMediaPermissionAdapter? camera;
  final RtcMediaPermissionAdapter? microphone;
  final RtcDeepLinkAdapter? deepLinks;
  final Future<void> Function()? pushNotifications;
  final RtcSecureStorageAdapter? secureStorage;
}

RtcHostAdapters? _activeHostAdapters;

RtcHostAdapters registerHostAdapters() {
  _activeHostAdapters ??= const RtcHostAdapters(
    camera: null,
    microphone: null,
    deepLinks: null,
    pushNotifications: null,
    secureStorage: null,
  );
  return _activeHostAdapters!;
}

RtcHostAdapters getHostAdapters() => registerHostAdapters();

File: apps/sdkwork-rtc-pc/src/bootstrap/hostAdapters.ts
export interface RtcMediaPermissionAdapter {
  requestCamera(): Promise<boolean>;
  requestMicrophone(): Promise<boolean>;
}

export interface RtcDeepLinkAdapter {
  getCurrentPath(): string;
  subscribe(listener: (path: string) => void): () => void;
}

export interface RtcSecureStorageAdapter {
  getItem(key: string): string | null;
  setItem(key: string, value: string): void;
  removeItem(key: string): void;
}

export interface RtcHostAdapters {
  camera: RtcMediaPermissionAdapter | null;
  deepLinks: RtcDeepLinkAdapter | null;
  pushNotifications: null;
  secureStorage: RtcSecureStorageAdapter | null;
}

let activeHostAdapters: RtcHostAdapters | null = null;

function createBrowserHostAdapters(): RtcHostAdapters {
  const secureStorage: RtcSecureStorageAdapter | null =
    typeof window === "undefined"
      ? null
      : {
          getItem(key) {
            return window.sessionStorage.getItem(key);
          },
          setItem(key, value) {
            window.sessionStorage.setItem(key, value);
          },
          removeItem(key) {
            window.sessionStorage.removeItem(key);
          },
        };

  const camera: RtcMediaPermissionAdapter | null =
    typeof navigator === "undefined" || !navigator.mediaDevices?.getUserMedia
      ? null
      : {
          async requestCamera() {
            try {
              const stream = await navigator.mediaDevices.getUserMedia({ video: true });
              stream.getTracks().forEach((track) => track.stop());
              return true;
            } catch {
              return false;
            }
          },
          async requestMicrophone() {
            try {
              const stream = await navigator.mediaDevices.getUserMedia({ audio: true });
              stream.getTracks().forEach((track) => track.stop());
              return true;
            } catch {
              return false;
            }
          },
        };

  const deepLinks: RtcDeepLinkAdapter | null =
    typeof window === "undefined"
      ? null
      : {
          getCurrentPath() {
            return `${window.location.pathname}${window.location.hash}`;
          },
          subscribe(listener) {
            const handleChange = () => listener(`${window.location.pathname}${window.location.hash}`);
            window.addEventListener("hashchange", handleChange);
            window.addEventListener("popstate", handleChange);
            return () => {
              window.removeEventListener("hashchange", handleChange);
              window.removeEventListener("popstate", handleChange);
            };
          },
        };

  return {
    camera,
    deepLinks,
    pushNotifications: null,
    secureStorage,
  };
}

export function registerHostAdapters(): RtcHostAdapters {
  if (!activeHostAdapters) {
    activeHostAdapters = createBrowserHostAdapters();
  }
  return activeHostAdapters;
}

export function getHostAdapters(): RtcHostAdapters {
  return activeHostAdapters ?? registerHostAdapters();
}
